### PR TITLE
Archaeologist basic core

### PR DESCRIPTION
**Pull request by wmarkowitz**
Currently DECAY needs to be implemented on the frontend. DECAY properly works on the backend in RunGym and the like, but whenever DECAY triggers in the client, the game ends with a defeat immediately.

~~**NOT YET READY TO BE MERGED**~~

### CODE REVIEW

File: cards/src/main/resources/cards/custom/group8/occultist/classic/spell_aid_of_the_ancient.json [hunk 1]
```diff
@@ -1,18 +1,28 @@
...
+         "class": "SummonSpell",
+         "cards": [
+           "token_tauntentacle",
+           "token_tauntentacle",
+           "token_tauntentacle"
+         ]
+       },
+       {
+         "class": "DamageSpell",
+         "target": "FRIENDLY_HERO",
```
Comment (doctorpangloss) [inline cards/src/main/resources/cards/custom/group8/occultist/classic/spell_aid_of_the_ancient.json]: this should have `ignoreSpellDamage`

File: cards/src/main/resources/cards/custom/group8/occultist/basic/spell_madman_knowledge.json [hunk 2]
```diff
@@ -4,23 +4,19 @@
...
-  "targetSelection": "FRIENDLY_MINIONS",
+  "description": "Take 2 damage. Draw a card",
+  "targetSelection": "NONE",
   "spell": {
     "class": "MetaSpell",
     "spells": [
       {
-        "class": "DestroySpell"
+        "class": "DamageSpell",
+        "target": "FRIENDLY_HERO",
```
Comment (doctorpangloss) [inline cards/src/main/resources/cards/custom/group8/occultist/basic/spell_madman_knowledge.json]: needs `ignoreSpellDamage`

File: cards/src/main/resources/cards/custom/group8/occultist/tgt/spell_bloody_blow.json [hunk 3]
```diff
@@ -1,28 +1,18 @@
...
-  "rarity": "COMMON",
-  "description": "Take $2 damage. Deal $2 damage to up to two random enemy minions.",
-  "condition": {
-    "class": "MinionCountCondition",
-    "value": 0,
-    "operation": "GREATER",
-    "targetPlayer": "OPPONENT"
-  },
+  "rarity": "FREE",
+  "description": "Take $2 damage. Deal $1 damage to all minions.",
```
Comment (doctorpangloss) [inline cards/src/main/resources/cards/custom/group8/occultist/tgt/spell_bloody_blow.json]: `Take 2 damage`

i think this effect is meant to ignore spell damage

File: cards/src/main/resources/cards/custom/group8/occultist/tgt/spell_bloody_blow.json [hunk 4]
```diff
@@ -1,28 +1,18 @@
...
-        "spell": {
-          "class": "DamageSpell",
-          "value": 2
-        }
+        "class": "DamageSpell",
+        "target": "ALL_MINIONS",
+        "value": 1
       },
       {
         "class": "DamageSpell",
```
Comment (doctorpangloss) [inline cards/src/main/resources/cards/custom/group8/occultist/tgt/spell_bloody_blow.json]: `ignoreSpellDamage`

File: game/src/main/java/net/demilich/metastone/game/spells/AddDescriptionSpell.java [hunk 5]
```diff
@@ -22,17 +23,24 @@
...
+			if (dynamicDescription != null && dynamicDescription.length > 0) {
+				target.setAttribute(Attribute.DYNAMIC_DESCRIPTION,
+						Stream.concat(
+								Arrays.stream((DynamicDescriptionDesc[]) target.getAttributes().getOrDefault(Attribute.DYNAMIC_DESCRIPTION,
+										new DynamicDescriptionDesc[0])), Arrays.stream(dynamicDescription)).toArray(DynamicDescriptionDesc[]::new));
+			}
+		} else if (desc.containsKey(SpellArg.DESCRIPTION)) {
+			description = desc.getString(SpellArg.DESCRIPTION);
+		}
+		if (target.getDescription().equals("")) {
```
Comment (doctorpangloss) [inline game/src/main/java/net/demilich/metastone/game/spells/AddDescriptionSpell.java]: should probably be `Objects.equals` (target or getDescription can be null)

File: game/src/main/java/net/demilich/metastone/game/spells/BuffAsEnchantmentSpell.java [hunk 6]
```diff
@@ -45,22 +45,26 @@ protected void onCast(GameContext context, Player player, SpellDesc desc, Entity
...
-		logger.debug("onCast {} {}: {} gains ({})", context.getGameId(), source, target, attackBonus + "/" + (hpBonus + armorBonus));
+		logger.debug("onCast {} {}: {} gains ({})", context.getGameId(), source, target, attackBonus + "/" + (hpBonus + totalArmorBonus));
 
 		if (target instanceof Hero) {
-			if (armorBonus != 0) {
-				context.getLogic().gainArmor(context.getPlayer(target.getOwner()), armorBonus);
+			if (totalArmorBonus != 0) {
+				context.getLogic().gainArmor(context.getPlayer(target.getOwner()), totalArmorBonus);
+			}
+			if (decayingArmorBonus != 0) {
```
Comment (doctorpangloss) [inline game/src/main/java/net/demilich/metastone/game/spells/BuffAsEnchantmentSpell.java]: it was discussed to remove decaying armor

File: game/src/main/java/net/demilich/metastone/game/spells/BuffAsEnchantmentSpell.java [hunk 7]
```diff
@@ -45,22 +45,26 @@ protected void onCast(GameContext context, Player player, SpellDesc desc, Entity
 		checkArguments(logger, context, source, desc, SpellArg.ATTACK_BONUS, SpellArg.HP_BONUS, SpellArg.ARMOR_BONUS, SpellArg.VALUE);
 		int attackBonus = desc.getValue(SpellArg.ATTACK_BONUS, context, player, target, source, 0);
 		int hpBonus = desc.getValue(SpellArg.HP_BONUS, context, player, target, source, 0);
-		int armorBonus = desc.getValue(SpellArg.ARMOR_BONUS, context, player, target, source, 0);
+		int decayingArmorBonus = desc.getValue(SpellArg.DECAYING_ARMOR_BONUS, context, player, target, source, 0);
```
Comment (doctorpangloss) [inline game/src/main/java/net/demilich/metastone/game/spells/BuffAsEnchantmentSpell.java]: remove ...

File: game/src/main/java/net/demilich/metastone/game/spells/BuffSpell.java [hunk 8]
```diff
@@ -124,7 +124,8 @@ protected SpellDesc getReverseSpell(GameContext context, Player player, Entity s
 		if (desc.get(SpellArg.VALUE) instanceof ValueProvider
 				|| desc.get(SpellArg.ATTACK_BONUS) instanceof ValueProvider
 				|| desc.get(SpellArg.HP_BONUS) instanceof ValueProvider
-				|| desc.get(SpellArg.ARMOR_BONUS) instanceof ValueProvider) {
+				|| desc.get(SpellArg.ARMOR_BONUS) instanceof ValueProvider
+				|| desc.get(SpellArg.DECAYING_ARMOR_BONUS) instanceof ValueProvider) {
```
Comment (doctorpangloss) [inline game/src/main/java/net/demilich/metastone/game/spells/BuffSpell.java]: remove ...

File: game/src/main/java/net/demilich/metastone/game/spells/desc/SpellArg.java [hunk 9]
```diff
@@ -89,6 +89,10 @@
 	 * @see net.demilich.metastone.game.spells.BuffSpell for an example of a spell that uses this arg.
 	 */
 	ARMOR_BONUS,
+	/**
+	 * The amount of decaying armor to buff to an entity, which is almost always only a hero.
+	 */
+	DECAYING_ARMOR_BONUS,
```
Comment (doctorpangloss) [inline game/src/main/java/net/demilich/metastone/game/spells/desc/SpellArg.java]: bye bye

File: testutils/src/main/resources/test/spell_test_give_decay.json [hunk 10]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "Give Decay",
+  "baseManaCost": 0,
+  "type": "SPELL",
+  "heroClass": "ANY",
+  "rarity": "FREE",
+  "description": "Give a minion Taunt.",
```
Comment (doctorpangloss) [inline testutils/src/main/resources/test/spell_test_give_decay.json]: should probably say Decay

File: cards/src/main/resources/cards/custom/arch/misc/spell_blast_apart.json [hunk 11]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "Blast Apart",
+  "baseManaCost": 9,
+  "type": "SPELL",
+  "heroClass": "LIGHTBROWN",
+  "rarity": "EPIC",
+  "description": "Destroy a minion and deal its Health as damage to all other minions.",
```
[the remark this answers is not in the document]
Comment (wmarkowitz) [reply cards/src/main/resources/cards/custom/arch/misc/spell_blast_apart.json]: this shares wording with Shadowflame, so this seems fine.

https://hearthstone.gamepedia.com/Shadowflame

**Comment by wmarkowitz** [comment]
I ran the tests locally, and the failing test is the one for Signs of the End, which @doombubbles already seems to have already put in a fix for in #1308

**Pull request merged**